src/Spago/Command/Fetch.purs

```diff
@@ -404,42 +404,45 @@ getTransitiveDepsFromPackageSet packageSet deps = do
+      if Set.member dep seen then do
+        State.modify_ $ cycleError dep
       else do
-        cache <- State.get
-        case Map.lookup dep cache of
-          Just allDeps ->
-            pure (init { packages = allDeps })
-          Nothing ->
-            -- First look for the package in the set to get a version number out,
-            -- then use that version to look it up in the index and get the dependencies
-            case Map.lookup dep packageSet of
-              Nothing -> pure (init { errors { notInPackageSet = Set.singleton dep } })
-              Just package -> do
-                maybeDeps <- State.lift $ memoisedGetPackageDependencies dep package
-                case maybeDeps of
-                  Nothing -> pure (init { errors { notInIndex = Set.singleton dep } })
-                  Just dependenciesMap -> do
-                    -- recur here, as we need to get the transitive tree, not just the first level
-                    { packages: childDeps, errors } <-
-                      for (Map.toUnfoldable dependenciesMap :: Array (Tuple PackageName Range)) (\(Tuple d _) -> go (Set.insert dep seen) d)
-                        >>= (pure <<< foldl mergeResults init)
-                    let allDeps = Map.insert dep package childDeps
-                    when (Set.isEmpty (errors.cycle <> errors.notInIndex <> errors.notInPackageSet)) do
-                      State.modify_ $ Map.insert dep allDeps
-                    pure { packages: allDeps, errors }
+        -- If the package is a transitive dependency of some other package that
+        -- we already met, then we don't need to do the work again
+        alreadyRun <- Map.member dep <$> State.gets _.packages
+        when (not alreadyRun)
+          -- If we need to compute the dependencies from scratch instead, we first look
+          -- in the package set to get a version number out, then use that version to
+          -- look it up in the index and get the dependencies
+          case Map.lookup dep packageSet of
+            Nothing -> State.modify_ $ notInPackageSetError dep
+            Just package -> do
+              maybeDeps <- State.lift $ memoisedGetPackageDependencies dep package
+              case maybeDeps of
+                Nothing -> State.modify_ $ notInIndexError dep
+                Just dependenciesMap -> do
+                  -- Compare errors before and after recursively running transitive deps
+                  errors <- State.gets _.errors
+
+                  -- recur here, as we need to get the transitive tree, not just the first level
+                  void $ forWithIndex dependenciesMap
+                    (\dependency _ -> go (Set.insert dep seen) dependency)
+
+                  -- Errors may have changed after running through the child deps
+                  errorsAfterTransitiveDeps <- State.gets _.errors
+
+                  -- Do not include the package if any child deps fail
```

Comment:
This makes me wonder: are we repeating work unnecessarily? If we have errors while resolving dependencies for the package A then we don't include it in the state, but then the next time (in another dependency branch) that we encounter it then we will try to resolve these again?

Reply:
It's necessary for the circular dependencies test. The expectation in the test is for both package a and package b to be included in the error, without this condition only one of them is.

Reply:
I think I was referring to something else, but I reread the implementation a couple more times and I'm convinced it is correct 😄